def process_arguments(args, sequences, output_dir):
    arguments = {"seqids": sequences,
                "database": args.database,
                "input_file": None if sequences == [] else args.input.split("/")[-1],
                "input_file_db_const": args.input_seqs_db_const,
                "consensus": args.consensus,
                "KEGG_ID": args.kegg,
                "InterPro_ID": args.interpro,
                "hmm_database_name": args.hmm_db_name,
                "alignment_method": args.align_method.lower(),
                "msa_aligner": args.aligner,
                "input_type": None if args.input == None else args.input_type,
                "metagenomic": True if args.input_type == "metagenome" else False,
                "hmm_validation": args.hmm_validation,
                "expansion": args.expansion,
                "concat_models": args.concat_hmm_models,
                "output_directory": output_dir,
                "out_table_format": args.output_type,
                "hmmsearch_out_type": args.hmms_output_type,
                "threads": args.threads,
                "workflow": args.workflow,
                "thresholds": [*range(60, 91, 5)] if args.expansion else False,
                "verbose": args.verbose,
                "overwrite": args.overwrite
                }
    return arguments
